guides/python-testing.md

```diff
@@ -0,0 +1,138 @@
+# Testing Python Packages
+
+The following guide is to layout a general guideline for testing Python
+packages. This document is not to be followed exactly word-by-word, but rather
+it lays out multiple things you should consider. Feel free to adopt and adapt
+as you deem necessary.
+
+## Do I need to test my Python package?
+
+Yes. Next question.
+
+## What should I use to run my tests?
+
+[pytest](https://docs.pytest.org/en/latest/) is recommended.
+There are many ``pytest`` plugins that you could utilize to get the most
+out of your tests. Example of several useful ones:
+
+* [pytest-doctestplus](https://github.com/astropy/pytest-doctestplus) for
+  advanced features for testing example code in documentation.
+* [pytest-openfiles](https://github.com/astropy/pytest-openfiles) for
+  detecting file handles that were inadvertently left open at the end of
+  unit tests.
+* [pytest-arraydiff](https://github.com/astrofrog/pytest-arraydiff) for
+  generation and comparison of data arrays produced during unit tests.
+* [pytest-socket](https://github.com/miketheman/pytest-socket) for handling
+  network calls.
+* [pytest-faulthandler](https://github.com/pytest-dev/pytest-faulthandler)
+  for getting a useful traceback out of segmentation fault.
+* [ci-watson](https://github.com/spacetelescope/ci_watson) for functionality
+  related to STScI Jenkins and Artifactory usage.
+
+## Where should I put my tests?
+
+See the templates in Python Packaging section.
+
+Template aside, there is also the question of whether you want to bundle the
+tests with your package or not. While bundling allows users to test their
+installations, it also increases the package size. For more information,
+see "Choosing a test layout / import rules" in
+https://docs.pytest.org/en/latest/goodpractices.html .
+
+## What should I test?
+
+In theory, there are many kinds of software testing. For instance,
+unit testing, integration testing, regression testing, etc. Currently, this
+document only focuses on unit and regression tests, which are usually
```

Comment:
You note here that you will only cover unit and regression tests, but then there's a bit at the end talking about CI testing....

Reply:
I am still not sure where is the boundary between regression and CI. In my experience, they both go hand in hand. Please suggest better verbiage if possible. 😅

Reply:
Maybe something like this...

"Currently, this document focuses on unit and regression tests (usually "shipped" with the package they are tied to), as well as continuous integration testing.